File: apps/fluster/lib/features/command_palette/state/actions/append_command_palette_category.dart
import 'package:fluster/core/state/global_state.dart';
import 'package:fluster/core/types/state_types.dart';
import 'package:fluster/features/command_palette/data/models/command_palette_category.dart';

class AppendCommandPaletteCategoryAction extends FlusterAction {
  final CommandPaletteCategory cat;
  AppendCommandPaletteCategoryAction(this.cat);

  @override
  GlobalAppState reduce() {
    return state.copyWith(
      commandPaletteState: state.commandPaletteState.copyWith(
        navigationStack: [...state.commandPaletteState.navigationStack, cat],
        filteredItems: cat.items,
        open: true,
        selectedIndex: 0,
      ),
    );
  }
}

File: apps/fluster/src/features/command_palette/state/actions/appendCommandPaletteCategory.ts
import { CommandPaletteCategory } from "#/command_palette/data/models/command_palette_category";
import { Location } from "react-router";
import {
    CommandPaletteActionType,
    CommandPaletteContextActions,
} from "../command_palette_provider";
import { Dispatch } from "react";

export const appendCommandPaletteCategory = async (
    cat: CommandPaletteCategory,
    location: Location,
    dispatch: Dispatch<CommandPaletteContextActions>
): Promise<void> => {
    const items = await cat.getItems(location);
    dispatch({
        type: CommandPaletteActionType.appendCommandPaletteCategory,
        payload: {
            cat: cat,
            items: items,
        },
    });
};
